fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/IndexingMutuallyByRecords.java

```diff
@@ -0,0 +1,529 @@
+                LogMessageKeys.INDEXING_FRAGMENTATION_FIRST, fragmentFirst,
+                LogMessageKeys.INDEXING_FRAGMENTATION_CURRENT, fragmentCurrent,
+                LogMessageKeys.INDEXING_FRAGMENTATION_TYPE, fragmentIterationType));
+    }
+
+    private Range fragmentGet() {
+        int i = getFragmentCurrent();
+        // Note that fragmentNum is fragmentBoundaries.size()-1
+        Tuple low = fragmentBoundaries.get(i);
+        Tuple high = fragmentBoundaries.get(i + 1);
+        byte[] lowBytes = low == null ? new byte[]{(byte)0} : low.pack();
+        byte[] highBytes = high == null ? new byte[]{(byte)0xff} : high.pack();
+        return new Range(lowBytes, highBytes);
+    }
+
+    private int getFragmentCurrent() {
+        return fragmentCurrent;
+    }
+
+    private void fragmentPlusPlus() {
+        fragmentCurrent += fragmentStep;
+        fragmentCurrent %= fragmentNum;
+        if (getFragmentCurrent() == fragmentFirst) {
+            fragmentIterationTypePlusPlus();
+        }
+    }
+
+    private void fragmentIterationTypePlusPlus() {
+        if (fragmentIterationType == FragmentIterationType.ANY) {
+            fragmentIterationType = FragmentIterationType.RECOVER;
+        }
+        if (fragmentIterationType == FragmentIterationType.FULL) {
+            fragmentIterationType = FragmentIterationType.ANY;
+        }
+    }
+
+    @Nonnull
+    @Override
+    CompletableFuture<Void> buildIndexInternalAsync() {
+        return getRunner().runAsync(context -> openRecordStore(context)
+                .thenCompose( store -> context.getReadVersionAsync()
+                        .thenCompose(vignore -> {
+                            // init fragment data
+                            setFragmentationData(store);
+                            SubspaceProvider subspaceProvider = common.getRecordStoreBuilder().getSubspaceProvider();
+                            // validation checks, if any, will be performed here
+                            return subspaceProvider.getSubspaceAsync(context)
+                                    .thenCompose(subspace -> buildMultiTargetIndex(subspaceProvider, subspace));
+                        })
+                ), common.indexLogMessageKeyValues("IndexingMutuallyByRecords::buildIndexInternalAsync"));
+    }
+
+    @Nonnull
+    private CompletableFuture<Void> buildMultiTargetIndex(@Nonnull SubspaceProvider subspaceProvider, @Nonnull Subspace subspace) {
+        final TupleRange tupleRange = common.computeRecordsRange();
+        final byte[] rangeStart;
+        final byte[] rangeEnd;
+        if (tupleRange == null) {
+            rangeStart = rangeEnd = null;
+        } else {
+            final Range range = tupleRange.toRange();
+            rangeStart = range.begin;
+            rangeEnd = range.end;
+        }
+
+        final CompletableFuture<FDBRecordStore> maybePresetRangeFuture =
+                rangeStart == null ?
```

Comment:
Is it possible for `rangeStart` to be non-null, but `rangeEnd` to be `null`?

Reply:
IIUC, not with the current `TupleRange` implementation: 
```
 return new Range(lowBytes == null ? new byte[0] : lowBytes, highBytes == null ? new byte[]{(byte)0xff} : highBytes);
```